### PR TITLE
String concat operator

### PR DESCRIPTION
**Pull request by adityagoel4512**
### Motivation and Context
Closes https://github.com/microsoft/onnxruntime/issues/17595.

### CODE REVIEW

File: cmake/deps.txt
Comment (adityagoel4512) [reply cmake/deps.txt:19]: This was because the new ONNX release had not yet been integrated into ORT. I will undraft and remove these changes now.

**Comment by adityagoel4512** [comment]
@xadupre would it be possible to get a review here. Also https://github.com/microsoft/onnxruntime/pull/18016, https://github.com/microsoft/onnxruntime/pull/18002 and https://github.com/microsoft/onnxruntime/pull/17977 are outstanding if someone could review these.

File: onnxruntime/test/providers/cpu/nn/stringconcat_test.cc
Comment (xadupre) [inline onnxruntime/test/providers/cpu/nn/stringconcat_test.cc:20]: Can you split the line into multiple lines? clang-format -i <filename> shoud work.

File: onnxruntime/core/providers/cpu/nn/string_concat.cc
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/nn/string_concat.cc:21]: This  is not performant due to lots of re-allocations and copies.
I suggest doing broadcast separate, compute the length of the resulting string, pre-allocate space and append to it.

**Review by yuslepukhin — CHANGES_REQUESTED** [review]

File: onnxruntime/core/providers/cpu/nn/string_concat.cc
Comment (gramalingam) [reply onnxruntime/core/providers/cpu/nn/string_concat.cc:21]: I am not sure I understand the suggestion. I understand the issue of how best to concatenate the two strings into a target string (to avoid redundant temporaries/copies etc.). But I am unclear about your suggestion for doing something during broadcast, and something later, for parallelization.

File: onnxruntime/test/providers/cpu/nn/stringconcat_test.cc
Comment (edgchen1) [inline onnxruntime/test/providers/cpu/nn/stringconcat_test.cc:1]: nit: name the file string_concat_test.cc for consistency with string_concat.h/cc

File: onnxruntime/core/providers/cpu/nn/string_concat.h
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/nn/string_concat.h:1]: general: add file copyright headers

File: onnxruntime/test/providers/cpu/nn/stringconcat_test.cc
Comment (edgchen1) [inline onnxruntime/test/providers/cpu/nn/stringconcat_test.cc:25]: maybe add a test where the first input is broadcast?

File: onnxruntime/core/providers/cpu/nn/string_concat.cc
Comment (adityagoel4512) [reply onnxruntime/core/providers/cpu/nn/string_concat.cc]: I would guess that @yuslepukhin would prefer to avoid any memory allocation when doing the actual string concatenation in parallel, but could you clarify this @yuslepukhin ?

Comment (gramalingam) [reply onnxruntime/core/providers/cpu/nn/string_concat.cc:21]: But the memory-allocation needs to happen at some point? We can minimize it (to one allocation), but cannot avoid it entirely, right?

Comment (yuslepukhin) [reply onnxruntime/core/providers/cpu/nn/string_concat.cc]: Disregard this comment.

**Comment by thiagocrepaldi** [comment]
@adityagoel4512 any update on this? ORT 1.17 release is coming soon :)

**Comment by adityagoel4512** [comment]
> @adityagoel4512 any update on this? ORT 1.17 release is coming soon :)

Waiting on a review round/CI pipelines to be run by a maintainer.

File: onnxruntime/test/providers/cpu/nn/stringconcat_test.cc
Comment (yuslepukhin) [reply onnxruntime/test/providers/cpu/nn/stringconcat_test.cc]: Per coding standards line length limit is 120.

Comment (adityagoel4512) [reply onnxruntime/test/providers/cpu/nn/stringconcat_test.cc]: This has been addressed at the latest stage of the PR

**Review by yuslepukhin — APPROVED** [review]

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Comment by yuslepukhin** [comment]
Please, check Generate documentation pipeline. It requires docs edits.

**Comment by yuslepukhin** [comment]
Just realized that you may not be able to see all output. Thank you for your diligence. One more step is required is to adjust documentation:
[StringConcatKernel.txt](https://github.com/microsoft/onnxruntime/files/13893012/StringConcatKernel.txt)

**Pull request merged**
